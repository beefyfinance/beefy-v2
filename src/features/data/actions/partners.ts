import { createAsyncThunk } from '@reduxjs/toolkit';
import { PartnersConfig } from '../apis/config';
import { getConfigApi } from '../apis/instances';

export const fetchPartnersConfig = createAsyncThunk<PartnersConfig>(
  'vaults/fetchPartnersConfig',
  async () => {
    const api = getConfigApi();
    return api.fetchPartnersConfig();
  }
);
