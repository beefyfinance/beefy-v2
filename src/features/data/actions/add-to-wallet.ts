import type { ChainEntity } from '../entities/chain';
import { createAsyncThunk } from '@reduxjs/toolkit';
import type { BeefyState } from '../../../redux-types';
import type { TokenEntity } from '../entities/token';
import { selectTokenByAddress } from '../selectors/tokens';
import { getSingleAssetSrc } from '../../../helpers/singleAssetSrc';

export type AddToWalletParams = {
  chainId: ChainEntity['id'];
  tokenAddress: TokenEntity['address'];
  customIconUrl?: string;
};

export type AddTokenToWalletPayload = {
  token: TokenEntity;
  iconUrl: string;
};

function getTokenIconUrl(
  id: TokenEntity['id'],
  chainId: TokenEntity['chainId'],
  customUrl?: string
): string {
  if (customUrl) {
    return customUrl;
  }

  const assetSrc = getSingleAssetSrc(id, chainId);
  if (assetSrc) {
    return `${window.location.origin}${assetSrc}`;
  }

  return '';
}

export const addTokenToWalletAction = createAsyncThunk<
  AddTokenToWalletPayload,
  AddToWalletParams,
  { state: BeefyState }
>('addToWallet/open', async ({ chainId, tokenAddress, customIconUrl }, { getState }) => {
  const state = getState();
  const token = selectTokenByAddress(state, chainId, tokenAddress);
  const iconUrl = getTokenIconUrl(token.id, token.chainId, customIconUrl);

  return {
    token,
    iconUrl,
  };
});
