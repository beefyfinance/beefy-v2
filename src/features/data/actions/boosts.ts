import { createAsyncThunk } from '@reduxjs/toolkit';
import type { BeefyState } from '../../../redux-types';
import type { TokenAllowance } from '../apis/allowance/allowance-types';
import type { FetchAllBalancesResult } from '../apis/balance/balance-types';
import { getAllowanceApi, getBalanceApi, getConfigApi } from '../apis/instances';
import type { BoostEntity } from '../entities/boost';
import type { ChainEntity } from '../entities/chain';
import { selectBoostById } from '../selectors/boosts';
import { selectChainById } from '../selectors/chains';
import { selectErc20TokenByAddress } from '../selectors/tokens';
import { selectVaultById } from '../selectors/vaults';
import type { BoostCampaignConfig, BoostConfig, BoostPartnerConfig } from '../apis/config-types';

// given the list of vaults is pulled from some api at some point
// we use the api to create an action
// this action should return just enough data for the state to work with

export interface FulfilledAllBoostsPayload {
  boostsByChainId: Record<ChainEntity['id'], BoostConfig[]>;
  partnersById: Record<string, BoostPartnerConfig>;
  campaignsById: Record<string, BoostCampaignConfig>;
}

export const fetchAllBoosts = createAsyncThunk<FulfilledAllBoostsPayload>(
  'boosts/fetchAllBoosts',
  async () => {
    const api = await getConfigApi();
    return api.fetchAllBoosts();
  }
);

interface InitBoostFormParams {
  boostId: BoostEntity['id'];
  mode: 'stake' | 'unstake';
  walletAddress: string | undefined;
}

interface InitBoostFormPayload {
  boostId: BoostEntity['id'];
  mode: 'stake' | 'unstake';

  // really, this should be separated
  walletAddress: string | undefined;
  balance: FetchAllBalancesResult;
  allowance: TokenAllowance[];

  // reducers below need to access the state
  state: BeefyState;
}

export const initiateBoostForm = createAsyncThunk<
  InitBoostFormPayload,
  InitBoostFormParams,
  { state: BeefyState }
>('boosts/initBoostForm', async ({ boostId, mode, walletAddress }, { getState }) => {
  const boost = selectBoostById(getState(), boostId);
  const vault = selectVaultById(getState(), boost.vaultId);
  const chain = selectChainById(getState(), boost.chainId);

  const balanceApi = await getBalanceApi(chain);

  const balanceRes: FetchAllBalancesResult = walletAddress
    ? await balanceApi.fetchAllBalances(getState(), [], [], [boost], walletAddress)
    : { tokens: [], boosts: [], govVaults: [] };

  const spenderAddress = boost.earnContractAddress;

  const allowanceApi = await getAllowanceApi(chain);
  const mooToken = selectErc20TokenByAddress(
    getState(),
    boost.chainId,
    vault.earnedTokenAddress,
    false
  );
  const allowanceRes =
    walletAddress && spenderAddress
      ? await allowanceApi.fetchTokensAllowance(
          getState(),
          [mooToken],
          walletAddress,
          spenderAddress
        )
      : [];

  return {
    boostId,
    mode,
    walletAddress,
    allowance: allowanceRes,
    balance: balanceRes,
    state: getState(),
  };
});
