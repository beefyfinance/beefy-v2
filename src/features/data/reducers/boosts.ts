import { createSlice } from '@reduxjs/toolkit';
import { WritableDraft } from 'immer/dist/internal';
import { fetchAllBoosts } from '../actions/boosts';
import { BoostConfig } from '../apis/config';
import { BoostEntity, isBoostActive, isBoostPreStake } from '../entities/boost';
import { ChainEntity } from '../entities/chain';
import { VaultEntity } from '../entities/vault';
import { NormalizedEntity } from '../utils/normalized-entity';

/**
 * State containing Vault infos
 */
export type BoostsState = NormalizedEntity<BoostEntity> & {
  byVaultId: {
    [vaultId: VaultEntity['id']]: {
      allBoostsIds: BoostEntity['id'][];
      prestakeBoostsIds: BoostEntity['id'][];
      activeBoostsIds: BoostEntity['id'][];
    };
  };
  byChainId: {
    [chainId: ChainEntity['id']]: {
      allBoostsIds: BoostEntity['id'][];
      prestakeBoostsIds: BoostEntity['id'][];
      activeBoostsIds: BoostEntity['id'][];
    };
  };
};
export const initialBoostsState: BoostsState = {
  byId: {},
  allIds: [],
  byVaultId: {},
  byChainId: {},
};

export const boostsSlice = createSlice({
  name: 'boosts',
  initialState: initialBoostsState,
  reducers: {
    // standard reducer logic, with auto-generated action types per reducer
  },
  extraReducers: builder => {
    // when boost list is fetched, add all new tokens
    builder.addCase(fetchAllBoosts.fulfilled, (sliceState, action) => {
      for (const [chainId, boosts] of Object.entries(action.payload)) {
        for (const boost of boosts) {
          addBoostToState(sliceState, chainId, boost);
        }
      }
    });
  },
});

function addBoostToState(
  sliceState: WritableDraft<BoostsState>,
  chainId: ChainEntity['id'],
  apiBoost: BoostConfig
) {
  if (apiBoost.id in sliceState.byId) {
    return;
  }

  const boost: BoostEntity = {
    id: apiBoost.id,
    chainId: chainId,
    assets: apiBoost.assets,
    earnedTokenId: apiBoost.earnedOracleId,
    earnContractAddress: apiBoost.earnContractAddress,
    logo: apiBoost.logo,
    name: apiBoost.name,
    partnerIds: apiBoost.partners ? apiBoost.partners.map(p => p.website) : [],
    status: apiBoost.status as BoostEntity['status'],
    vaultId: apiBoost.poolId,
  };
  sliceState.byId[boost.id] = boost;
  sliceState.allIds.push(boost.id);

  // add to vault id index
  if (sliceState.byVaultId[boost.vaultId] === undefined) {
    sliceState.byVaultId[boost.vaultId] = {
      allBoostsIds: [],
      activeBoostsIds: [],
      prestakeBoostsIds: [],
    };
  }
  sliceState.byVaultId[boost.vaultId].allBoostsIds.push(boost.id);
  if (isBoostActive(boost)) {
    sliceState.byVaultId[boost.vaultId].activeBoostsIds.push(boost.id);
  }
  if (isBoostPreStake(boost)) {
    sliceState.byVaultId[boost.vaultId].prestakeBoostsIds.push(boost.id);
  }

  // add to chain id index
  if (sliceState.byChainId[chainId] === undefined) {
    sliceState.byChainId[chainId] = {
      allBoostsIds: [],
      activeBoostsIds: [],
      prestakeBoostsIds: [],
    };
  }
  sliceState.byChainId[chainId].allBoostsIds.push(boost.id);
  if (isBoostActive(boost)) {
    sliceState.byChainId[chainId].activeBoostsIds.push(boost.id);
  }
  if (isBoostPreStake(boost)) {
    sliceState.byChainId[chainId].prestakeBoostsIds.push(boost.id);
  }
}
