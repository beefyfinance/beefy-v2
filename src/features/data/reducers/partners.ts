import { createSlice } from '@reduxjs/toolkit';
import { fetchAllBoosts } from '../actions/boosts';
import { fetchPartnersConfig } from '../actions/partners';
import { MoonpotConfig } from '../apis/config';
import { ChainEntity } from '../entities/chain';
import { PartnerEntity } from '../entities/partner';
import { VaultEntity } from '../entities/vault';

/**
 * State containing Vault infos
 */
export type PartnersState = {
  byId: {
    [partnerId: PartnerEntity['id']]: PartnerEntity;
  };
  moonpot: {
    byVaultId: {
      [vaultId: VaultEntity['id']]: MoonpotConfig;
    };
  };
  insurace: {
    byChainId: {
      [chainId: ChainEntity['id']]: boolean;
    };
  };
  qidao: {
    byVaultId: {
      [vaultId: VaultEntity['id']]: boolean;
    };
  };
  lacucina: {
    byVaultId: {
      [vaultId: VaultEntity['id']]: boolean;
    };
  };
};
export const initialPartnersState: PartnersState = {
  byId: {},
  moonpot: {
    byVaultId: {},
  },
  insurace: {
    byChainId: {},
  },
  qidao: {
    byVaultId: {},
  },
  lacucina: {
    byVaultId: {},
  },
};

export const partnersSlice = createSlice({
  name: 'partners',
  initialState: initialPartnersState,
  reducers: {
    // standard reducer logic, with auto-generated action types per reducer
  },
  extraReducers: builder => {
    builder.addCase(fetchPartnersConfig.fulfilled, (sliceState, action) => {
      for (const moonpotConfig of action.payload.Moonpot) {
        const vaultId = moonpotConfig.id;
        if (!sliceState.moonpot.byVaultId[vaultId]) {
          sliceState.moonpot.byVaultId[vaultId] = moonpotConfig;
        }
      }
      for (const chainId of action.payload.Insurace) {
        if (!sliceState.insurace.byChainId[chainId]) {
          sliceState.insurace.byChainId[chainId] = true;
        }
      }
      for (const vaultId of action.payload.QiDao) {
        if (!sliceState.qidao.byVaultId[vaultId]) {
          sliceState.qidao.byVaultId[vaultId] = true;
        }
      }
      for (const vaultId of action.payload.LaCucina) {
        if (!sliceState.lacucina.byVaultId[vaultId]) {
          sliceState.lacucina.byVaultId[vaultId] = true;
        }
      }
    });

    builder.addCase(fetchAllBoosts.fulfilled, (sliceState, action) => {
      for (const boosts of Object.values(action.payload)) {
        for (const boost of boosts) {
          for (const partner of boost.partners || []) {
            const id = partner.website;
            if (sliceState.byId[id] === undefined) {
              sliceState.byId[id] = { id, ...partner };
            }
          }
        }
      }
    });
  },
});
