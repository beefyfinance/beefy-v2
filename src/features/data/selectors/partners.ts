import { BeefyState } from '../../../redux-types';
import { VaultEntity } from '../entities/vault';
import { selectActiveVaultBoostIds, selectBoostById } from './boosts';
import { selectVaultById } from './vaults';

export const selectIsVaultMoonpot = (state: BeefyState, vaultId: VaultEntity['id']) => {
  return state.biz.partners.moonpot.byVaultId[vaultId] !== undefined;
};
export const selectMoonpotData = (state: BeefyState, vaultId: VaultEntity['id']) => {
  const mpd = state.biz.partners.moonpot.byVaultId[vaultId];
  if (!mpd) {
    throw new Error(`Vault ${vaultId} do not have moonpot data`);
  }
  return mpd;
};
export const selectIsVaultQidao = (state: BeefyState, vaultId: VaultEntity['id']) => {
  return state.biz.partners.qidao.byVaultId[vaultId] !== undefined;
};
export const selectIsVaultLacucina = (state: BeefyState, vaultId: VaultEntity['id']) => {
  return state.biz.partners.lacucina.byVaultId[vaultId] !== undefined;
};
export const selectIsVaultInsurace = (state: BeefyState, vaultId: VaultEntity['id']) => {
  const vault = selectVaultById(state, vaultId);
  return state.biz.partners.insurace.byChainId[vault.chainId] !== undefined;
};
export const selectIsVaultBinSpirit = (state: BeefyState, vaultId: VaultEntity['id']) => {
  const vault = selectVaultById(state, vaultId);
  return vault.id === 'beefy-binspirit';
};

export const selectBoostedVaultMainPartner = (state: BeefyState, vaultId: VaultEntity['id']) => {
  const boostIds = selectActiveVaultBoostIds(state, vaultId);
  const boost = selectBoostById(state, boostIds[0]);
  const partner = state.biz.partners.byId[boost.partnerIds[0]];
  return partner;
};
