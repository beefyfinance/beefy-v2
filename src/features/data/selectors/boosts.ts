import type { BeefyState } from '../../../redux-types';
import type { BoostEntity } from '../entities/boost';
import type { ChainEntity } from '../entities/chain';
import type { VaultEntity } from '../entities/vault';
import { getBoostStatusFromContractState } from '../reducers/boosts';
import { selectBoostUserBalanceInToken, selectBoostUserRewardsInToken } from './balance';
import { createCachedSelector } from 're-reselect';
import { BIG_ZERO } from '../../../helpers/big-number';

export const selectBoostById = createCachedSelector(
  (state: BeefyState) => state.entities.boosts.byId,
  (state: BeefyState, boostId: BoostEntity['id']) => boostId,
  (boostsById, boostId) => {
    if (boostsById[boostId] === undefined) {
      throw new Error(`selectBoostById: Unknown boost id ${boostId}`);
    }
    return boostsById[boostId];
  }
)((state: BeefyState, boostId: BoostEntity['id']) => boostId);

export const selectIsBoostActive = (state: BeefyState, boostId: BoostEntity['id']) => {
  const status = getBoostStatusFromContractState(boostId, selectBoostContractState(state, boostId));
  return status === 'active';
};

export const selectIsBoostActiveOrPreStake = (state: BeefyState, boostId: BoostEntity['id']) => {
  const status = getBoostStatusFromContractState(boostId, selectBoostContractState(state, boostId));
  return status === 'active' || status === 'prestake';
};

export const selectBoostsByChainId = createCachedSelector(
  (state: BeefyState, chainId: ChainEntity['id']) =>
    state.entities.boosts.byChainId[chainId]?.allBoostsIds,
  boostsByChain => boostsByChain || []
)((state: BeefyState, chainId: ChainEntity['id']) => chainId);

export const selectIsVaultBoosted = createCachedSelector(
  (state: BeefyState, vaultId: VaultEntity['id']) => selectActiveVaultBoostIds(state, vaultId),
  activeBoostIds => activeBoostIds.length > 0
)((state: BeefyState, vaultId: VaultEntity['id']) => vaultId);

export const selectIsVaultPreStakedOrBoosted = createCachedSelector(
  (state: BeefyState, vaultId: VaultEntity['id']) => selectActiveVaultBoostIds(state, vaultId),
  (state: BeefyState, vaultId: VaultEntity['id']) => selectPreStakeVaultBoostIds(state, vaultId),
  (activeBoostIds, prestakeBoostIds) => activeBoostIds.length + prestakeBoostIds.length > 0
)((state: BeefyState, vaultId: VaultEntity['id']) => vaultId);

export const selectVaultCurrentBoostIdWithStatus = createCachedSelector(
  (state: BeefyState, vaultId: VaultEntity['id']) => selectActiveVaultBoostIds(state, vaultId),
  (state: BeefyState, vaultId: VaultEntity['id']) => selectPreStakeVaultBoostIds(state, vaultId),
  (activeBoostIds, prestakeBoostIds) => {
    if (activeBoostIds.length > 0) {
      return { id: activeBoostIds[0], status: 'active' };
    }
    if (prestakeBoostIds.length > 0) {
      return { id: prestakeBoostIds[0], status: 'prestake' };
    }
    return null;
  }
)((state: BeefyState, vaultId: VaultEntity['id']) => vaultId);

export const selectIsVaultPrestakedBoost = createCachedSelector(
  (state: BeefyState, vaultId: VaultEntity['id']) => selectPreStakeVaultBoostIds(state, vaultId),
  prestakeBoostIds => prestakeBoostIds.length > 0
)((state: BeefyState, vaultId: VaultEntity['id']) => vaultId);

export const selectActiveVaultBoostIds = createCachedSelector(
  (state: BeefyState, vaultId: VaultEntity['id']) =>
    state.entities.boosts.byVaultId[vaultId]?.activeBoostsIds,
  boostIds => boostIds || []
)((state: BeefyState, vaultId: VaultEntity['id']) => vaultId);

export const selectPreStakeVaultBoostIds = createCachedSelector(
  (state: BeefyState, vaultId: VaultEntity['id']) =>
    state.entities.boosts.byVaultId[vaultId]?.prestakeBoostsIds,
  boostIds => boostIds || []
)((state: BeefyState, vaultId: VaultEntity['id']) => vaultId);

export const selectPreStakeOrActiveBoostIds = createCachedSelector(
  (state: BeefyState, vaultId: VaultEntity['id']) => selectActiveVaultBoostIds(state, vaultId),
  (state: BeefyState, vaultId: VaultEntity['id']) => selectPreStakeVaultBoostIds(state, vaultId),
  (activeBoostIds, prestakeBoostIds) => activeBoostIds.concat(prestakeBoostIds)
)((state: BeefyState, vaultId: VaultEntity['id']) => vaultId);

export const selectAllVaultBoostIds = createCachedSelector(
  (state: BeefyState, vaultId: VaultEntity['id']) =>
    state.entities.boosts.byVaultId[vaultId]?.allBoostsIds,
  boostIds => boostIds || []
)((state: BeefyState, vaultId: VaultEntity['id']) => vaultId);

export const selectPastBoostIdsWithUserBalance = (
  state: BeefyState,
  vaultId: VaultEntity['id']
) => {
  if (!state.entities.boosts.byVaultId[vaultId]) {
    return [];
  }

  const vaultBoosts = state.entities.boosts.byVaultId[vaultId];

  const boostIds = [];
  for (const eolBoostId of vaultBoosts.expiredBoostsIds) {
    const userBalance = selectBoostUserBalanceInToken(state, eolBoostId);
    if (userBalance.gt(0)) {
      boostIds.push(eolBoostId);
      continue;
    }
    const userRewards = selectBoostUserRewardsInToken(state, eolBoostId);
    if (userRewards.gt(0)) {
      boostIds.push(eolBoostId);
    }
  }
  return boostIds;
};

export const selectShouldDisplayVaultBoost = (state: BeefyState, vaultId: VaultEntity['id']) => {
  if (!state.entities.boosts.byVaultId[vaultId]) {
    return false;
  }

  const vaultBoosts = state.entities.boosts.byVaultId[vaultId];

  // has an active boost
  if (vaultBoosts.activeBoostsIds.length > 0) {
    return true;
  }
  // has a prestaking boost
  if (vaultBoosts.prestakeBoostsIds.length > 0) {
    return true;
  }

  // OR, there was an ended boost and user staked into it
  if (selectPastBoostIdsWithUserBalance(state, vaultId).length > 0) {
    return true;
  }

  return false;
};

export const selectVaultsActiveBoostPeriodFinish = (
  state: BeefyState,
  vaultId: BoostEntity['id']
) => {
  const activeBoost = selectVaultCurrentBoostIdWithStatus(state, vaultId);
  return activeBoost ? selectBoostPeriodFinish(state, activeBoost.id) : null;
};

export const selectBoostPeriodFinish = (state: BeefyState, boostId: BoostEntity['id']) => {
  return state.entities.boosts.contractState[boostId]?.periodFinish || null;
};

export const selectBoostContractState = (state: BeefyState, boostId: BoostEntity['id']) => {
  return state.entities.boosts.contractState[boostId] || { periodFinish: null, isPreStake: true };
};

export const selectUserBalanceOnActiveOrPastBoost = (
  state: BeefyState,
  vaultId: VaultEntity['id']
) => {
  const isBoosted = selectIsVaultPreStakedOrBoosted(state, vaultId);

  const activeBoost = isBoosted
    ? selectBoostById(state, selectPreStakeOrActiveBoostIds(state, vaultId)[0])
    : null;

  const pastBoostsWithUserBalance = selectPastBoostIdsWithUserBalance(state, vaultId).map(
    boostId => {
      return boostId;
    }
  );

  const isBoostedOrHaveBalanceInPastBoost = isBoosted || pastBoostsWithUserBalance.length === 1;

  const boost = isBoosted
    ? selectBoostById(state, activeBoost.id)
    : pastBoostsWithUserBalance.length === 1
    ? selectBoostById(state, pastBoostsWithUserBalance[0])
    : null;

  const boostBalance = isBoostedOrHaveBalanceInPastBoost
    ? selectBoostUserBalanceInToken(state, boost.id)
    : BIG_ZERO;

  return boostBalance;
};

export const selectUserActiveBoostBalanceInToken = (
  state: BeefyState,
  vaultId: VaultEntity['id'],
  walletAddress?: string
) => {
  const isBoosted = selectIsVaultBoosted(state, vaultId);

  const activeBoost = isBoosted
    ? selectBoostById(state, selectPreStakeOrActiveBoostIds(state, vaultId)[0])
    : null;

  return isBoosted ? selectBoostUserBalanceInToken(state, activeBoost.id, walletAddress) : BIG_ZERO;
};
