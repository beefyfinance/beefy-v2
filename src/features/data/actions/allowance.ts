import { createAsyncThunk } from '@reduxjs/toolkit';
import type { BeefyState } from '../../../redux-types';
import type { FetchAllAllowanceResult } from '../apis/allowance/allowance-types';
import { getAllowanceApi } from '../apis/instances';
import type { ChainEntity } from '../entities/chain';
import type { TokenErc20 } from '../entities/token';
import type { VaultGov, VaultStandard } from '../entities/vault';
import { isGovVault } from '../entities/vault';
import { selectBoostById, selectBoostsByChainId } from '../selectors/boosts';
import { selectChainById } from '../selectors/chains';
import { selectVaultByChainId, selectVaultById } from '../selectors/vaults';
import { selectWalletAddress } from '../selectors/wallet';

interface ActionParams {
  chainId: ChainEntity['id'];
  walletAddress: string;
}

export interface FetchAllAllowanceFulfilledPayload {
  chainId: ChainEntity['id'];
  data: FetchAllAllowanceResult;
}

export const fetchAllAllowanceAction = createAsyncThunk<
  FetchAllAllowanceFulfilledPayload,
  ActionParams,
  { state: BeefyState }
>('allowance/fetchAllAllowanceAction', async ({ chainId, walletAddress }, { getState }) => {
  const state = getState();

  const userAddress = walletAddress || selectWalletAddress(state);
  const chain = selectChainById(state, chainId);
  const api = await getAllowanceApi(chain);

  // maybe have a way to retrieve those easily
  const boosts = selectBoostsByChainId(state, chainId).map(vaultId =>
    selectBoostById(state, vaultId)
  );
  const allVaults = selectVaultByChainId(state, chainId).map(vaultId =>
    selectVaultById(state, vaultId)
  );
  const standardVaults: VaultStandard[] = [];
  const govVaults: VaultGov[] = [];
  for (const vault of allVaults) {
    if (isGovVault(vault)) {
      govVaults.push(vault);
    } else {
      standardVaults.push(vault);
    }
  }

  // always re-fetch state as late as possible
  const data = await api.fetchAllAllowances(
    // always re-fetch state as late as possible
    getState(),
    standardVaults,
    govVaults,
    boosts,
    userAddress
  );
  return { chainId, data };
});

interface FetchAllowanceActionParams {
  chainId: ChainEntity['id'];
  spenderAddress: string;
  tokens: TokenErc20[];
  walletAddress: string;
}

export const fetchAllowanceAction = createAsyncThunk<
  FetchAllAllowanceFulfilledPayload,
  FetchAllowanceActionParams,
  { state: BeefyState }
>(
  'allowance/fetchAllowanceAction',
  async ({ chainId, spenderAddress, tokens, walletAddress }, { getState }) => {
    const state = getState();
    const userAddress = walletAddress || selectWalletAddress(state);
    const chain = selectChainById(state, chainId);
    const api = await getAllowanceApi(chain);

    const allowanceRes =
      userAddress && spenderAddress
        ? await api.fetchTokensAllowance(getState(), tokens, userAddress, spenderAddress)
        : [];

    return { chainId, data: allowanceRes };
  }
);
