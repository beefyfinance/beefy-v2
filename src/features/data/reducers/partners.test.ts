import { fetchAllBoosts, FulfilledAllBoostsPayload } from '../actions/boosts';
import { initialPartnersState, partnersSlice } from './partners';

describe('Partners slice tests', () => {
  it('should update state on fulfilled boosts list', () => {
    const payload: FulfilledAllBoostsPayload = {
      harmony: [
        // one with a partnership
        {
          id: 'moo_banana-pera',
          poolId: 'banana-banana-eol',
          name: 'Pera',
          logo: 'degens/BANANA.svg',
          earnedToken: 'PERA',
          earnedTokenDecimals: 18,
          earnedTokenAddress: '0xb9D8592E16A9c1a3AE6021CDDb324EaC1Cbc70d6',
          earnContractAddress: '0x38247fCE28480A7BEF2CB7aD134ce091Bd2E1a82',
          earnedOracle: 'tokens',
          earnedOracleId: 'PERA',
          partnership: true,
          status: 'active',
          isMooStaked: true,
          partners: [
            {
              logo: 'stake/pera/logo.png',
              background: 'stake/pera/bg.png',
              text: "Pera Finance is a decentralized trading competition platform with multi-layered yield farming opportunities. Traders, liquidity providers and holders yield farm together through the DeFi's first decentralized trading competition. Every on-chain PERA transaction (transfer, trade, or liquidity addition/ removal) generates a 2 % transaction fee.",
              website: 'https://pera.finance/',
              social: {
                telegram: 'https://t.me/perafinance',
                twitter: 'https://twitter.com/perafinance',
              },
            },
          ],
        },
      ],
    };
    const action = { type: fetchAllBoosts.fulfilled, payload: payload };
    const state = partnersSlice.reducer(initialPartnersState, action);
    expect(state).toMatchSnapshot();

    // getting the same vaults don't update the state object
    const beforeReDispatch = state.byId;
    const newState = partnersSlice.reducer(state, action);
    const afterReDispatch = newState.byId;
    expect(beforeReDispatch).toBe(afterReDispatch);
  });
});
