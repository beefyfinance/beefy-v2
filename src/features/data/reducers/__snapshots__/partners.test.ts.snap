// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Partners slice tests should update state on fulfilled boosts list 1`] = `
Object {
  "byId": Object {
    "https://pera.finance/": Object {
      "background": "stake/pera/bg.png",
      "id": "https://pera.finance/",
      "logo": "stake/pera/logo.png",
      "social": Object {
        "telegram": "https://t.me/perafinance",
        "twitter": "https://twitter.com/perafinance",
      },
      "text": "Pera Finance is a decentralized trading competition platform with multi-layered yield farming opportunities. Traders, liquidity providers and holders yield farm together through the DeFi's first decentralized trading competition. Every on-chain PERA transaction (transfer, trade, or liquidity addition/ removal) generates a 2 % transaction fee.",
      "website": "https://pera.finance/",
    },
  },
  "insurace": Object {
    "byChainId": Object {},
  },
  "lacucina": Object {
    "byVaultId": Object {},
  },
  "moonpot": Object {
    "byVaultId": Object {},
  },
  "qidao": Object {
    "byVaultId": Object {},
  },
}
`;
